module/os/freebsd/zfs/vdev_geom.c

```diff
@@ -1123,14 +1199,35 @@ vdev_geom_io_start(zio_t *zio)
 		zio->io_target_timestamp = zio_handle_io_delay(zio);
 		bp->bio_offset = zio->io_offset;
 		bp->bio_length = zio->io_size;
-		if (zio->io_type == ZIO_TYPE_READ) {
+		if (zio->io_type == ZIO_TYPE_READ)
 			bp->bio_cmd = BIO_READ;
-			bp->bio_data =
-			    abd_borrow_buf(zio->io_abd, zio->io_size);
-		} else {
+		else
 			bp->bio_cmd = BIO_WRITE;
-			bp->bio_data =
-			    abd_borrow_buf_copy(zio->io_abd, zio->io_size);
+
+		/*
+		 * If possible, represent scattered and/or gang ABD buffer to
+		 * GEOM as an array of physical pages.  It allows to satisfy
+		 * requirement of virtually contiguous buffer without copying.
+		 */
+		int pgs = vdev_geom_check_unmapped(zio, cp);
+		bp->bio_cflags = (pgs > 0);
+		if (pgs > 0) {
+			bp->bio_ma = malloc(sizeof (struct vm_page *) * pgs,
```

Comment:
What code would free the memory for bio_ma?

Reply:
See the vdev_geom_io_done() in next chunk.

Reply:
> That said, what is the point of filling bio_ma, if the buffer is already mapped?

If the buffer is mapped contiguously (abd_is_linear() returns true), then pgs == 0 and it should go traditional way without using bio_ma.  If it is scattered, then pgs > 0 and bio_ma is used.